crates/ruff_python_formatter/resources/test/fixtures/ruff/docstring_code_examples.py

```diff
@@ -0,0 +1,315 @@
+    """
+    pass
+
+
+# Tests that a blank PS2 line at the end of a doctest can get dropped.
+# It is treated as part of the Python snippet which will trim the
+# trailing whitespace.
+def doctest_blank_end():
+    """
+    Do cool stuff.
+
+    >>> def cool_stuff ( x ):
+    ...     print( x )
+    ...     print( x )
+    ...
+    """
+    pass
+
+
+# Tests that a blank PS2 line at the end of a doctest can get dropped
+# even when there is text following it.
+def doctest_blank_end_then_some_text():
+    """
+    Do cool stuff.
+
+    >>> def cool_stuff ( x ):
+    ...     print( x )
+    ...     print( x )
+    ...
+
+    And say something else.
+    """
+    pass
+
+
+# Test that a doctest containing a triple quoted string gets formatted
+# correctly and doesn't result in invalid syntax.
+def doctest_with_triple_single():
+    """
+    Do cool stuff.
+
+    >>> x        =      '''tricksy'''
+    """
+    pass
+
+
+# Test that a doctest containing a triple quoted f-string gets
+# formatted correctly and doesn't result in invalid syntax.
+def doctest_with_triple_single():
+    """
+    Do cool stuff.
+
+    >>> x    = f'''tricksy'''
+    """
+    pass
+
+
+# Another nested multi-line string case, but with triple escaped double
+# quotes inside a triple single quoted string.
+def doctest_with_triple_escaped_double():
+    """
+    Do cool stuff.
+
+    >>> x       =       '''\"\"\"'''
+    """
+    pass
+
+
+# Tests that inverting the triple quoting works as expected.
+def doctest_with_triple_inverted():
+    '''
+    Do cool stuff.
+
+    >>> x   =   """tricksy"""
+    '''
+    pass
+
+
+# Tests that inverting the triple quoting with an f-string works as
+# expected.
+def doctest_with_triple_inverted_fstring():
+    '''
+    Do cool stuff.
+
+    >>> x        = f"""tricksy"""
+    '''
+    pass
+
+
+# Tests nested doctests are ignored. That is, we don't format doctests
+# recursively. We only recognize "top level" doctests.
+#
+# This restriction primarily exists to avoid needing to deal with
+# nesting quotes. It also seems like a generally sensible restriction,
+# although it could be lifted if necessary I believe.
+def doctest_nested_doctest_not_formatted():
+    '''
+    Do cool stuff.
+
+    >>> def nested( x   ):
+    ...     """
+    ...     Do nested cool stuff.
+    ...     >>> func_call( 5 )
+    ...     """
+    ...     pass
+    '''
+    pass
+
+
+# Tests that the starting column does not matter.
+def doctest_varying_start_column():
+    '''
+    Do cool stuff.
+
+    >>> assert    ("Easy!")
+      >>> import                      math
+          >>> math.floor(  1.9  )
+          1
+    '''
+    pass
+
+
+# Tests that long lines get wrapped... appropriately.
+#
+# The docstring code formatter uses the same line width settings as for
+# formatting other code. This means that a line in the docstring can
+# actually extend past the configured line limit.
+#
+# It's not quite clear whether this is desirable or not. We could in
```

Comment:
Yeah I'm not sure. I suspect the behavior you have here _is_ desirable? (Any idea what other tools do?)

Reply:
So it looks like `blacken-docs` behaves like us. But [`blackdoc` passes the current indentation width to `black`](https://github.com/keewis/blackdoc/blob/822385ba8955e44397288caa07a2d2e3dc933b1e/blackdoc/blacken.py#L72) when reformatting a code snippet. From playing with it, it does indeed look like that the max line width is preserved uniformly.

I will also say that as someone who cares a lot about line width, I would want the formatter to wrap lines according to my setting in a global way. i.e., The code snippets shouldn't be treated as starting from the first column.

I'll take a quick look and see if I can make this work _robustly_. (Emphasis on "robust." This was why I didn't explore it too much initially.)

Reply:
A separate setting _might_ make sense e.g. `docstring_code_line_width`. I can see arguments for both sides. For me, I care more about a reasonable width in my reference documentation generated from my docstrings and would be okay with starting on the first column.

On a slightly separate note, `docstring_code_...` settings suggest that `docstring_code_enabled` may be a better name for toggling this behavior?

Reply:
Yeah, that's a good point.

Reply:
OK, here is what I suggest:

* We keep the current behavior which treats the code snippet as starting at column 1. This I think aligns with what @zanieb said about prioritizing the display of the code snippet.
* We create an issue tracking a new feature for overriding this behavior to do something different.

Probably `docstring_code_line_width` would be a good thing to add, but I think there's also a different mode of "respect the top-level line width setting." I don't think you can get that behavior by specifying a distinct concrete docstring code line width, since the formatter needs to take indentation into account.

So basically, there's some implementation work here to change the behavior and also some decisions that need to be made about how we present this behavior.

If this sounds good to y'all I can create a new issue capturing this.

Reply:
So one thing I uncovered today is that by allowing docstring lines to exceed the globally configured line width, you wind up with linter errors on long lines. For example, after running the docstring code snippet formatter on polars: https://github.com/BurntSushi/polars/actions/runs/7006619426/job/19058868021?pr=1

Reply:
Issue created: https://github.com/astral-sh/ruff/issues/8855